### DISCUSSION TITLE
Building dynamic peripheral

### DISCUSSION THREAD

**Discussion by SirEdvin** [cc-tweaked/CC-Tweaked#1516] [Q&A] [answered]
So, I have a strange question, how do I suppose do build dynamic peripheral myself? My dynamic peripherals is usually collected from plugins and I extract methods from them.

In past I have [used](https://github.com/SirEdvin/Peripheralium/blob/790d37b05fb08164364da37ddb14993e07fd938d/projects/core/src/main/kotlin/site/siredvin/peripheralium/api/peripheral/IPeripheralPlugin.kt#L14) GENERATOR field, that was hidden internally, but now it is even more hidden. 

Do I need to just re-implement Generator by myself or there is an API that I missed?

**Comment by SquidDev** [comment, marked as answer]
> how do I suppose [something involving CC's internals]

Uh oh, you're on dangerous ground there! :p.

To actually answer your question, I did a bunch of refactoring in 910a63214e395ecae6993d8e0487384c725b3dd3 to make the method cache per-server, in prep for #1382. You now need to use `ServerContext.get(server).peripheralMethods()` to get access to a `MethodSupplier<PeripheralMethod>`, which can then be used to enumerate the methods on an object.

If it's any consolation, as of f5b16261cc0980cd73afc3707cfe33310d9af46d/CC:T 1.106.0, we do now publish the internal common jar, so at least this shouldn't require any multi-loader code.

**Reply by SirEdvin** [reply]
Thanks for such quick answer, I will try to do this.

And also thanks for common jar :)
